bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -91,87 +91,162 @@
-                    public Object addingService(ServiceReference reference) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
-                        return provider;
-                    }
+    /**
+     * Activates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     * @throws Exception
+     */
+    protected void activate(ComponentContext componentContext) throws Exception {
+        this.bundleContext = componentContext.getBundleContext();
+        this.ruleEngine = new RuleEngine(bundleContext, moduleTypeRegistry);
+        this.ruleEngine.setStatusInfoCallback(this);
+        super.activate(bundleContext);
+    }
 
-                    @Override
-                    public void modifiedService(ServiceReference reference, Object service) {
-                        TemplateProvider provider = (TemplateProvider) bc.getService(reference);
-                        templateUpdated(provider.getTemplates(null));
+    /**
+     * Deactivates this component. Called from DS.
+     *
+     * @param componentContext this component context.
+     */
+    protected void deactivate(ComponentContext componentContext) {
+        super.deactivate();
+        // removeManagedProvider(managedProvider);
+        ruleEngine.dispose();
+        ruleEngine = null;
+    }
 
-                    }
+    /**
+     * Bind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void setModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = moduleTypeRegistry;
+    }
 
-                    @Override
-                    public void removedService(ServiceReference reference, Object service) {
-                    }
+    /**
+     * Unbind the {@link ModuleTypeRegistry} service - called from DS.
+     *
+     * @param moduleTypeRegistry moduleTypeRegistry service.
+     */
+    protected void unsetModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
+        this.moduleTypeRegistry = null;
+    }
 
-                });
-        templateProviderTracker.open();
+    /**
+     * Bind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void setTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = templateRegistry;
+    }
 
+    /**
+     * Unbind the {@link TemplateRegistry} service - called from DS.
+     *
+     * @param templateRegistry templateRegistry service.
+     */
+    protected void unsetTemplateRegistry(TemplateRegistry templateRegistry) {
+        this.templateRegistry = null;
+    }
+
+    protected void setStorageService(StorageService storageService) {
+        // this.storageService = storageService;
```

Comment:
1) Ok, RuleRegistry can have a 0:1 dependency on ManagedRuleProvider.
2) It it is a static dependency, the RuleRegistry can be sure to have the ManagedRuleProvider injected before activate() is called - so it can deal with the priority.
3) "It starts RuleEngine." -> imho, it shouldn't. The RuleEngine could be a service on its own, having a dependency on the RuleRegistry.
4) Then it definitely should be its on DS component and not be started by the RuleRegistry.